velox/exec/ExchangeClient.cpp

```diff
@@ -170,74 +174,42 @@ void ExchangeClient::request(const RequestSpec& requestSpec) {
   }
 }
 
-int32_t ExchangeClient::countPendingSourcesLocked() {
-  int32_t numPending = 0;
-  for (auto& source : sources_) {
-    if (source->isRequestPendingLocked()) {
-      ++numPending;
-    }
-  }
-  return numPending;
-}
-
-int64_t ExchangeClient::getAveragePageSize() {
-  auto averagePageSize =
-      std::min<int64_t>(maxQueuedBytes_, queue_->averageReceivedPageBytes());
-  if (averagePageSize == 0) {
-    averagePageSize = 1 << 20; // 1 MB.
+std::vector<ExchangeClient::RequestSpec>
+ExchangeClient::pickSourcesToRequestLocked() {
+  if (closed_) {
+    return {};
   }
-
-  return averagePageSize;
-}
-
-int32_t ExchangeClient::getNumSourcesToRequestLocked(int64_t averagePageSize) {
-  // Figure out how many more 'averagePageSize' fit into 'maxQueuedBytes_'.
-  // Make sure to leave room for 'numPending' pages.
-  const auto numPending = countPendingSourcesLocked();
-
-  auto numToRequest = std::max<int32_t>(
-      1, (maxQueuedBytes_ - queue_->totalBytes()) / averagePageSize);
-  if (numToRequest <= numPending) {
-    return 0;
+  std::vector<RequestSpec> requestSpecs;
+  while (!emptySources_.empty()) {
+    auto& source = emptySources_.front();
+    bool ok = source->shouldRequestLocked();
+    VELOX_CHECK(ok);
+    requestSpecs.push_back({std::move(source), 0});
+    emptySources_.pop();
   }
-
-  return numToRequest - numPending;
-}
-
-void ExchangeClient::pickSourcesToRequestLocked(
-    RequestSpec& requestSpec,
-    int32_t numToRequest,
-    std::queue<std::shared_ptr<ExchangeSource>>& sources) {
-  while (requestSpec.sources.size() < numToRequest && !sources.empty()) {
-    auto& source = sources.front();
-    if (source->shouldRequestLocked()) {
-      requestSpec.sources.push_back(source);
+  int64_t availableSpace =
+      maxQueuedBytes_ - queue_->totalBytes() - totalPendingBytes_;
+  while (availableSpace > 0 && !producingSources_.empty()) {
+    auto& source = producingSources_.front();
+    int64_t requestBytes = 0;
+    for (auto bytes : source->remainingBytes()) {
+      availableSpace -= bytes;
+      if (availableSpace <= 0) {
```

Comment:
We may want to queue at least one source (even if it doesn't fit).

Reply:
Is this to break out the the case that even the one page could be larger than our budget?

Reply:
I think it is possible that we have no pending data fetches and min data size is > out budget. If this happens, we may get stuck, no?